File: go/paymentrouter/merger.go
package paymentrouter

import (
	"context"
	"net/http"
	"sync"
)

// MergeChallenges calls GetChallenge concurrently for every adapter that has
// an entry in routeCfg. Results are merged with http.Header.Add (not Set) so
// that multi-value headers (e.g. WWW-Authenticate) accumulate. A single
// failure invokes onError and is skipped; nil headers are skipped silently.
func MergeChallenges(
	ctx context.Context,
	adapters []ProtocolAdapter,
	r *http.Request,
	routeCfg RouteConfig,
	onError func(err error, protocol string),
) http.Header {
	// Filter to adapters that have config in this route.
	type result struct {
		h   http.Header
		err error
		a   ProtocolAdapter
	}

	filtered := make([]ProtocolAdapter, 0, len(adapters))
	for _, a := range adapters {
		if _, ok := routeCfg[a.Name()]; ok {
			filtered = append(filtered, a)
		}
	}

	results := make([]result, len(filtered))
	var wg sync.WaitGroup
	wg.Add(len(filtered))

	for i, a := range filtered {
		i, a := i, a
		go func() {
			defer wg.Done()
			cfg := routeCfg[a.Name()]
			h, err := a.GetChallenge(ctx, r, cfg)
			results[i] = result{h: h, err: err, a: a}
		}()
	}
	wg.Wait()

	merged := http.Header{}
	for _, res := range results {
		if res.err != nil {
			if onError != nil {
				onError(res.err, res.a.Name())
			}
			continue
		}
		if res.h == nil {
			continue
		}
		for key, vals := range res.h {
			for _, v := range vals {
				merged.Add(key, v)
			}
		}
	}
	return merged
}

File: typescript/bu-payments/app-payment-router/src/merger.ts
import type { ProtocolAdapter, UnifiedRouteConfig } from "./types.js";

/**
 * Concurrently collect challenge headers from every eligible adapter and
 * merge them into a single header map.
 *
 * - Only calls adapters that have an explicit entry in
 *   `routeConfig.adapterConfigs`.
 * - An adapter that throws is reported via `onError` and skipped; other
 *   adapters are unaffected.
 * - `WWW-Authenticate` values are concatenated with `", "`.
 * - For any other duplicate header, the later value wins.
 */
export async function mergeChallenges(
  request: Request,
  routeConfig: UnifiedRouteConfig,
  adapters: readonly ProtocolAdapter[],
  onError?: (error: unknown, protocol: string) => void,
): Promise<Record<string, string>> {
  const eligible = adapters.filter((a) => has(routeConfig, a.name));

  const results = await Promise.allSettled(
    eligible.map((a) =>
      a.buildChallenge(request, routeConfig.adapterConfigs[a.name]),
    ),
  );

  const merged: Record<string, string> = {};

  for (let i = 0; i < eligible.length; i++) {
    const r = results[i];
    const a = eligible[i];

    if (r.status === "rejected") {
      onError?.(r.reason, a.name);
      continue;
    }

    for (const [k, v] of Object.entries(r.value)) {
      if (k.toLowerCase() === "www-authenticate") {
        const existing = Object.keys(merged).find(
          (kk) => kk.toLowerCase() === "www-authenticate",
        );
        if (existing) {
          merged[existing] = `${merged[existing]}, ${v}`;
        } else {
          merged[k] = v;
        }
      } else {
        merged[k] = v;
      }
    }
  }

  return merged;
}

function has(cfg: UnifiedRouteConfig, name: string): boolean {
  return (
    cfg.adapterConfigs != null &&
    name in cfg.adapterConfigs &&
    cfg.adapterConfigs[name] != null
  );
}
